### ISSUE TITLE
Automatically translate a full third-party Rust crate

### ISSUE THREAD

**Issue by dudozermaks** [fzyzcjy/flutter_rust_bridge#1588]
Hi, thanks for the beautiful tool!
Is there any way I could port 1 crate without re-implementing all of it's members?

I already have a flutter project, just need to add rust crate to it.
I've tried:
1. clone [Sudoku](https://github.com/Emerentius/sudoku) crate (`git clone https://github.com/Emerentius/sudoku rust`)
2. `flutter_rust_bridge_codegen integrate`
3. adding all dependencies listed in [original](https://github.com/Emerentius/sudoku/blob/master/Cargo.toml) crate's `Cargo.toml`
4. `mv rust/src/lib.rs rust/src/api/main.rs` + adding it to `rust/src/api/mod.rs` like this: `pub mod main;` + remove the top line with `mod frb_generated`
5. created `rust/src/lib.rs` with this conent
```
pub mod api;
mod frb_generated; /* AUTO INJECTED BY flutter_rust_bridge. This line may not be accurate, and you can change it according to your needs. */
```
6. `flutter_rust_bridge_codegen generate`

And there is no file named `main.dart` in `lib/src/rust/api/` :(
Last command output is:
```
[0.3s] Parse
  └── [0.2s] Run cargo expand
  └── [0.1s] Parse source graph
[1.7s] Generate
  └── [0.1s] Run cbindgen
  └── [1.6s] Run ffigen
[0.4s] Polish
  └── [0.4s] Run Dart formatter
  └── [0.0s] Run Rust formatter                                                                     Done!
```
So there is no errors (probably...)

Since there is v2.0 release, I think it would be important to write a guide on how to port a crate (if it is possible). If we will figure it out I can do this.

And, I should mention, I'm a newbie in rust

**Comment by fzyzcjy** [comment]
Hi, do you mean porting something from v1 to v2 (https://cjycode.com/flutter_rust_bridge/guides/miscellaneous/upgrade), or mean using a crate directly without writing wrapper logic (#1587), or something else?

**Comment by dudozermaks** [comment]
Hi! I mean using crate without writing wrapper logic. In the issue you mentioned, one need to add crate as a dependency to `Cargo.toml`. That means that cargo will download this crate and store it somewhere. That means, that your tool can not access this code to generate bindings. I think, that if we have all of the crate code in the `rust` folder, then we can generate bindings automatically. Am I right?

**Comment by fzyzcjy** [comment]
> I think, that if we have all of the crate code in the rust folder, then we can generate bindings automatically. Am I right?

Yes, if the crate code has function signatures that flutter_rust_bridge supports. For example, lifetime specifiers are not supported, because there does not seem to be a way (feel free to suggest one!) to handle it in Dart. The famous pyo3 (python rust "bridge") also does not have lifetime support.

**Comment by dudozermaks** [comment]
Oh, I see. That's interesting...
I think there is two ways to implement that: by text and by code.

### By text
By text I mean, create some kind of documentation page, describing the process. for example, how to mark functions and structs to export (I'm not sure about this feature though, but it would be nice). This page should tell about:
* how to properly integrate. For example, after `flutter_rust_bridge_codegen integrate` `rust/Cargo.toml` would be removed and new would be emplaced. This means that one should copy `Cargo.toml` of the crate to the existing one.
* how to replace popular lifetimes. For example:`&str` to `String`.
* which files to add to the crate and how. For example `rust/src/api/some_file_with_some_bindigs.rs`

### By code
Create some kind of script that replaces lifetimes. But, to replace `&str` with `String` script also needs to add `.to_string()` at the end of the function. And if this one case seems pretty simple, imagine how much of this cases exists... If after all of this manipulations we somehow won't break internal crate logic, it would be a miracle. It would be nearly impossible to do that with a simple bash script. I'm not sure, if such tools, capable of replacing types, even exist. If we could somehow wrap lifetimes it would be much easier. But I don't see any way, at least on the Dart side of things.

**TL;DR;** Considering all of the above, I think most realistic option right now is to write guide on how to do that by hand.

**Comment by fzyzcjy** [comment]
For the "by code", I guess it is similar to, we just automatically generate a wrapper layer around it - but that still has the same problem as mentioned earlier.

As for "by text", I am not sure whether copy-and-paste-and-modify, which is suggested in the first post of this issue, is the best solution. Another way is to just depend on that crate, and write down some wrappers manually. This is how the minimal demo https://github.com/fzyzcjy/flutter_rust_bridge/blob/master/frb_example/gallery/rust/src/api/polars.rs does.

**Comment by JustSimplyKyle** [comment]
Question: aren't lifetimed references just types? Is there anything stopping us from just wrapping it in an opaque type?

**Comment by fzyzcjy** [comment]
Good question! I brainstormed and see one possible approach (not having time to think deeper, so this may or may not work!):

Recall that RustOpaque is Arc under the hood, so let's talk about Arc instead.
Suppose `fn f(a: &A) -> &B`. Then, we do not create `Arc<&B>` (which does not work IMHO), but create `Arc<(&B, Arc<A>)>)` (or sth like that). By doing so, we guarantee that `&B` lives shorter than `A`.
This may have other problems, such as circular reference, possibly unsafe code (I have not thought whether this can be avoided), etc...
